### ISSUE TITLE
Why am I seeing login and signup routes when I am authenticated?

### ISSUE THREAD

**Issue by aligajani** [sahat/satellizer#478]
Ideally, they shouldn't be shown as I am authenticated. I tried a fix using resolve on the state for login and signup routes, but that is not an extensible solution. Is there an abstract way where we can tell ui.router that these routes should only be accessible when the user is logged out. Something like that.

**Comment by sahat** [comment]
Don't know honestly, this should be asked on ui-router GitHub project. The best I can recommend is doing something like this, but you probably already know:

```js
resolve: {
  authenticated: function($q, $location, $auth) {
    var deferred = $q.defer();

    if ($auth.isAuthenticated()) {
      deferred.reject();
    } else {
      deferred.resolve();
    }

    return deferred.promise;
  }
}
```

**Comment by sahat** [comment]
I will add at least that resolve function to both login and signup ui-router states in the next update. It's still better than nothing.

**Comment by aligajani** [comment]
I have a new solution and I will push it to Satellizer. It uses events from
ui.router.

On Wed, Jul 15, 2015 at 2:39 AM, Sahat Yalkabov <notifications@github.com>
wrote:

> Reopened #478 <https://github.com/sahat/satellizer/issues/478>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sahat/satellizer/issues/478#event-355993255>.
>



-- 


Ali Gajani
Founder at Mr. Geek
www.mrgeek.me
www.aligajani.com

**Comment by aligajani** [comment]
@sahat Have a look at this. Uses `ui.router` events. Correct and scalable as compared to resolve.

https://gist.github.com/aligajani/9da7f8be799b2ef69be5

**Comment by jculverwell** [comment]
Hi Guys,

I've used the open source Auth0 library for AngularJS and they did something similar to below.  I've modified to work with Satellizer.

I like this approach as you simply need to mark your states with. 

data: {
        requiresLogin: true
        }

Hope this helps and thank you for a great library.

Jim

angular.module('MyApp', ['ngResource', 'ngMessages', 'ui.router', 'mgcrea.ngStrap', 'satellizer', 'server-validate'])
  .config(function($stateProvider, $urlRouterProvider, $authProvider) {
    $stateProvider
      .state('home', {
        url: '/',
        templateUrl: 'app/partials/home.html',
        data: {
        requiresLogin: true
        }
      })
      .state('login', {
        url: '/login',
        templateUrl: 'app/partials/login.html',
        controller: 'LoginCtrl'
      })
      .state('signup', {
        url: '/signup',
        templateUrl: 'app/partials/signup.html',
        controller: 'SignupCtrl'
      })
      .state('logout', {
        url: '/logout',
        template: null,
        controller: 'LogoutCtrl'
      })
      .state('profile', {
        url: '/profile',
        templateUrl: 'app/partials/profile.html',
        controller: 'ProfileCtrl',
        data: {
            requiresLogin: true
        }
      });

    $urlRouterProvider.otherwise('/');

  }).run(function ($rootScope, $auth, $state) {
      $rootScope.$on('$stateChangeStart', function (e, to) {
          if (to.data && to.data.requiresLogin) {
              if (!$auth.isAuthenticated()) {
                  e.preventDefault();
                  $state.go('login');
              }
          }
      });
  });;

**Comment by aligajani** [comment]
This is flawed. You can access /login when authenticated.

My solution is superior: https://gist.github.com/aligajani/9da7f8be799b2ef69be5

**Comment by jculverwell** [comment]
Your approach looks fine - I ended up with the approach below.  Basically I had a root abstract state which defined everything as protected.  I then overrode this for states which did not need to be protected.

.config(function($stateProvider, $urlRouterProvider, $authProvider) {
      $stateProvider
      .state('root', {
          url: '',
          abstract: true,
          views: {
              'header': {
                  templateUrl: 'app/partials/header.html',
                  controller: 'HeaderCtrl'
              },
              'footer': {
                  templateUrl: 'app/partials/footer.html',
                  //controller: 'FooterCtrl'
              }
          },
          data: {
              requiresLogin: true
          }
      })
      .state('root.home', {
          url: '/',
          views: {
              'container@': {
                  templateUrl: 'app/partials/home.html',
                  controller: 'HomeCtrl'
              }
          }
      })
      .state('root.profile', {
            url: '/profile',
            views: {
                'container@': {
                    templateUrl: 'app/partials/profile.html',
                    controller: 'ProfileCtrl'
                }
            }
       })
      .state('root.login', {
          url: '/login',
          views: {
              'container@': {
                  templateUrl: 'app/partials/login.html',
                  controller: 'LoginCtrl'
              }
          },
          data: {
              requiresLogin: false
          }
      })
     .state('root.signup', {
        url: '/signup',
        views: {
            'container@': {
                templateUrl: 'app/partials/signup.html',
                controller: 'SignUpCtrl'
            }
        },
        data: {
            requiresLogin: false
        }
     })
    .state('root.logout', {
        url: '/logout',
        views: {
            'container@': {
                template: null,
                controller: 'LogoutCtrl',
            }
        },
        data: {
            requiresLogin: false
        }
    });

      $urlRouterProvider.otherwise('/');

  }).run(function ($rootScope, $auth, $state) {
      $rootScope.$on('$stateChangeStart', function (e, to) {
          if (to.data && to.data.requiresLogin) {
              if (!$auth.isAuthenticated()) {
                  e.preventDefault();
                  $state.go('root.login');
              }
          }
      });
  });;

**Comment by aligajani** [comment]
Wouldn't this allow you to access /login even when authenticated?

On Thu, Jul 23, 2015 at 7:19 AM, Jim Culverwell <notifications@github.com>
wrote:

> Your approach looks fine - I ended up with the approach below. Basically I
> had a root abstract state which defined everything as protected. I then
> overrode this for states which did not need to be protected.
>
> .config(function($stateProvider, $urlRouterProvider, $authProvider) {
> $stateProvider
> .state('root', {
> url: '',
> abstract: true,
> views: {
> 'header': {
> templateUrl: 'app/partials/header.html',
> controller: 'HeaderCtrl'
> },
> 'footer': {
> templateUrl: 'app/partials/footer.html',
> //controller: 'FooterCtrl'
> }
> },
> data: {
> requiresLogin: true
> }
> })
> .state('root.home', {
> url: '/',
> views: {
> 'container@': {
> templateUrl: 'app/partials/home.html',
> controller: 'HomeCtrl'
> }
> }
> })
> .state('root.profile', {
> url: '/profile',
> views: {
> 'container@': {
> templateUrl: 'app/partials/profile.html',
> controller: 'ProfileCtrl'
> }
> }
> })
> .state('root.login', {
> url: '/login',
> views: {
> 'container@': {
> templateUrl: 'app/partials/login.html',
> controller: 'LoginCtrl'
> }
> },
> data: {
> requiresLogin: false
> }
> })
> .state('root.signup', {
> url: '/signup',
> views: {
> 'container@': {
> templateUrl: 'app/partials/signup.html',
> controller: 'SignUpCtrl'
> }
> },
> data: {
> requiresLogin: false
> }
> })
> .state('root.logout', {
> url: '/logout',
> views: {
> 'container@': {
> template: null,
> controller: 'LogoutCtrl',
> }
> },
> data: {
> requiresLogin: false
> }
> });
>
>   $urlRouterProvider.otherwise('/');
>
> }).run(function ($rootScope, $auth, $state) {
> $rootScope.$on('$stateChangeStart', function (e, to) {
> if (to.data && to.data.requiresLogin) {
> if (!$auth.isAuthenticated()) {
> e.preventDefault();
> $state.go('root.login');
> }
> }
> });
> });;
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sahat/satellizer/issues/478#issuecomment-123990891>.
>



-- 


Ali Gajani
Founder at Mr. Geek
www.mrgeek.me
www.aligajani.com

**Comment by jculverwell** [comment]
I don't know the correct answer to this, do you think this is an issue?
What you you think should happen when you try to access login but are
already authenticated?  By being able to access login doesn't it give you a
chance to login as a different user?

Cheers

Jim

**Comment by aligajani** [comment]
When you are authenticated, /login should redirect back you to the previous page. 

Like Redirect::back() in Laravel.

I implemented this which makes it work as aforementioned: https://gist.github.com/aligajani/9da7f8be799b2ef69be5

**Comment by sahat** [comment]
Resolved in [0.12](https://github.com/sahat/satellizer/releases/tag/0.12.0) release.
